common4j/src/main/com/microsoft/identity/common/java/cache/SharedPreferencesAccountCredentialCache.java

```diff
@@ -100,113 +107,150 @@ public SharedPreferencesAccountCredentialCache(
         Logger.verbose(TAG, "Init: " + TAG);
         mSharedPreferencesFileManager = sharedPreferencesFileManager;
         mCacheValueDelegate = accountCacheValueDelegate;
+        new Thread(() -> load()).start();
+    }
+
+    private void load()
+    {
+        final String methodTag = TAG + ":load";
+
+        synchronized (mCacheLock) {
+            try {
+                mCachedAccountRecordsWithKeys = loadAccountsWithKeys();
+                Logger.info(methodTag, "Loaded " + mCachedAccountRecordsWithKeys.size() + " AccountRecords");
+                mCachedCredentialsWithKeys = loadCredentialsWithKeys();
+                Logger.info(methodTag, "Loaded " + mCachedCredentialsWithKeys.size() + " Credentials");
+            } catch (Throwable t) {
+                Logger.error(methodTag, "Failed to load initial credentials from SharedPreferences", t);
+            } finally {
+                mLoaded = true;
+                mCacheLock.notifyAll();
+            }
+        }
+    }
+
+    private void waitForInitialLoad() {
+        while (!mLoaded) {
+            try {
+                mCacheLock.wait();
+            } catch (InterruptedException e) {}
+        }
     }
 
     @Override
-    public synchronized void saveAccount(@NonNull final AccountRecord accountToSave) {
-        Logger.verbose(TAG, "Saving Account...");
-        Logger.verbose(TAG, "Account type: [" + accountToSave.getClass().getSimpleName() + "]");
+    public void saveAccount(@NonNull final AccountRecord accountInput) {
+        final String methodTag = TAG + ":saveAccount";
+
+        AccountRecord accountToSave = null;
+        try {
+            accountToSave = (AccountRecord) accountInput.clone();
+        } catch (CloneNotSupportedException e) {
+            Logger.error(methodTag, "Failed to clone AccountRecord", e);
```

Comment:
So we are not going to save the account in this case? Should we just throw an Exception rather than returning silently out of this method?

Reply:
This exception needing to be handled at all is an oddity of the Java implementation of `clone()`.  Because we've implemented it on this base type, we can't actually end up here.  This 'throws' is still part of the method signature though, in case you've managed to somehow call this on an object that doesn't implement Cloneable but still exposed a public method called `clone()` that called into the protected `super.clone()`:
https://cs.android.com/android/platform/superproject/+/master:libcore/ojluni/src/main/java/java/lang/Object.java;l=261?q=object.clone&sq=

I don't think we should give the higher level code any more conditions to handle -- logging this impossibility here seems enough.